File: 46_Permutations/permute.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        def dfs(stack, nums):
            if not nums:
                res.append(stack)
                return

            for idx, num in enumerate(nums):
                dfs(stack + [num], nums[0:idx] + nums[idx + 1:])

        dfs([], nums)
        return res

sol = Solution()
nums = [1, 2]
print sol.permute(nums)

File: 46_Permutations/permute.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
    	vector<vector<int> > permute(vector<int>& nums) {
        	vector<vector<int> > res;
            dfs(res, nums, 0);
            return res;
    	}
    private:
        void dfs(vector<vector<int> > &res, vector<int> &nums, int begin) {
            if(begin >= nums.size()) {
                res.push_back(nums);
                return;
            }

            for(int i = begin; i < nums.size(); i++) {
                swap(nums[begin], nums[i]);
                dfs(res, nums, begin + 1);
                swap(nums[begin], nums[i]);
            }
        }
};
